### PR TITLE
fix(v2): render children in BrowserOnly after client is ready

### LINKED ISSUE #4934: BrowserOnly hydration issuse
### Prerequisites

- [x] I'm using the latest version of Docusaurus.
- [x] I have tried the `npm run clear` or `yarn clear` command.
- [x] I have tried `rm -rf node_modules yarn.lock package-lock.json` and re-installing packages.

### Description

BrowserOnly seems to have a hydration issue. It

- removes the first parent node of the would-be content
- and then picks the next sibling of the page and ... wrap the rest that is left in it

Say you have:

```jsx
<BrowserOnly>
  () => ..............
    <div class="a">
      <div class="b">hello</div>
    </div>
</BrowserOnly>
<div class="c">...</div>
```

In production (not in dev) Docusaurus will transform it to:

```jsx
<div class="c">
  <div class="b">hello</div>
</div>
<div class="c">...</div>
```

Here is a real world example:

Screenshot 2021-06-09 at 09 12 35

Docusaurus will start to mix that code in very strange ways:

Screenshot 2021-06-09 at 09 15 59

And this is how it looks in dev:

Screenshot 2021-06-09 at 09 17 18

A workaround is to wrap BrowserOnly into the the first element of the would-be content. Why this works is even stranger, i can't see a pattern in that. 

Screenshot 2021-06-09 at 09 26 43

### PR DESCRIPTION
**Pull request by lex111**
## Motivation

Fixes #4934

Currently child elements inside `BrowserOnly` are rendered too early, before the client side is ready, because in some cases the final markup may break.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

See To reproduce section in #4934 (eg. insert code below on index page).

```js
<BrowserOnly
  fallback={
    <div className="a">
      <div className="b">Will be render during SSR</div>
    </div>
  }>
  {() => (
    <div className="a">
      <div className="b">Will be render on client (i.e. replace SSR content above)</div>
    </div>
  )}
</BrowserOnly>
```

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/facebook/docusaurus, and link to your PR here.)

### CODE REVIEW

File: packages/docusaurus/src/client/exports/BrowserOnly.tsx [hunk 1]
```diff
@@ -15,11 +16,13 @@ function BrowserOnly({
   children?: () => JSX.Element;
   fallback?: JSX.Element;
 }): JSX.Element | null {
+  const {isClient} = useDocusaurusContext();
+
   if (!ExecutionEnvironment.canUseDOM || children == null) {
```
Comment (slorber) [inline packages/docusaurus/src/client/exports/BrowserOnly.tsx:21]: I'd rather use 

```
  if (!isClient || !ExecutionEnvironment.canUseDOM || children == null)
```

We need to have the exact same markup on server and on first client render, so this is not good enough, as it does not include the potential fallback (not always null): `isClient ? <>{children()}</> : null;`

**Review by slorber — CHANGES_REQUESTED** [review]
Thanks

> Oops, sorry, it is my mistake, will be fixed in #4935.

Not sure to understand, did it use to work and was changed recently? It looks like this problem is has potentially always been there

File: packages/docusaurus/src/client/exports/BrowserOnly.tsx (hunk 1 above)
Comment (lex111) [reply packages/docusaurus/src/client/exports/BrowserOnly.tsx]: Ah, yes, you are right, just fixed that.

**Comment by slorber** [comment]
LGTM

**Pull request merged**